core/iwasm/libraries/wasi-nn/src/wasi_nn_openvino.c

```diff
@@ -0,0 +1,537 @@
+/*
+ * Copyright (C) 2019 Intel Corporation.  All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include "wasi_nn_types.h"
+#include "wasi_nn_openvino.h"
+#include "logger.h"
+#include "bh_platform.h"
+
+#include "openvino/c/openvino.h"
+
+/*
+ * refer to
+ * https://docs.openvino.ai/2024/openvino-workflow/running-inference/integrate-openvino-with-your-application.html
+ *
+ * Steps about integrating OpenVINO are:
+ *
+ * 1. Create OpenVINO Runtime Core
+ * 2. Compile Model
+ * 3. Create Inference Request
+ * 4. Set Inputs
+ * 5. Start Inference
+ * 6. Process Inference Results
+ *
+ * from 4. to 6. is the Inference Loop
+ */
+
+typedef struct {
+    ov_core_t *core;
+    /* keep input model files */
+    void *weight_data;
+    ov_tensor_t *weights_tensor;
+    ov_model_t *model;
+    /* add prepostprocess */
+    ov_model_t *new_model;
+    ov_compiled_model_t *compiled_model;
+    ov_infer_request_t *infer_request;
+    ov_tensor_t *input_tensor;
+} OpenVINOContext;
+
+/*
+ * BE AWARE OF "goto fail"
+ */
+#define CHECK_OV_STATUS(status, error_code)                \
+    do {                                                   \
+        ov_status_e s = status;                            \
+        if (s != OK) {                                     \
+            NN_ERR_PRINTF("return status \"%s\", line %d", \
+                          ov_get_error_info(s), __LINE__); \
+            error_code = runtime_error;                    \
+            goto fail;                                     \
+        }                                                  \
+    } while (0)
+
+static void
+dump_ov_shape_t(const ov_shape_t *shape, int32_t output_len, char *output)
+{
+    int ret = 0;
+
+    ret = snprintf(output, output_len, "%ld,[", shape->rank);
+    if (!ret)
+        return;
+
+    output_len -= ret;
+    output += ret;
+
+    for (unsigned i = 0; i < shape->rank && output_len; i++) {
+        ret = snprintf(output, output_len, " %ld", shape->dims[i]);
+        if (!ret)
+            return;
+
+        output_len -= ret;
+        output += ret;
+    }
+
+    snprintf(output, output_len, "]");
+    return;
+}
+
+#ifndef NDEBUG
+static void
+print_model_input_output_info(ov_model_t *model)
+{
+    char *friendly_name = NULL;
+    ov_model_get_friendly_name(model, &friendly_name);
+    NN_DBG_PRINTF("model name: %s", friendly_name);
+    ov_free(friendly_name);
+
+    size_t input_size = 0;
+    ov_model_inputs_size(model, &input_size);
+    for (unsigned i = 0; i < input_size; i++) {
+        ov_output_const_port_t *input_port = NULL;
+        ov_shape_t input_shape = { 0 };
+        ov_element_type_e input_type;
+        char shape_info[64] = { 0 };
+
+        ov_model_const_input_by_index(model, i, &input_port);
+        ov_const_port_get_shape(input_port, &input_shape);
```

Comment:
Had better check returned input_port and input_shape since they will be freed later?